### ISSUE TITLE
Sometimes https://peterportal.org/ doesn't redirect to course search

### ISSUE THREAD

**Issue by js0mmer** [icssc/peterportal-client#277]
Occasionally when I go to [https://peterportal.org/](https://peterportal.org/) it doesn't redirect to [https://peterportal.org/search/courses](https://peterportal.org/search/courses) and I remain on a blank page. I am unable to reproduce it consistently and don't know why it happens but I've had it happen by chance on three devices now (laptop, PC, and phone). I am use the latest version of Chrome on all of them if that happens to be important.

Example on my phone:
https://user-images.githubusercontent.com/8922227/224222830-11f6e623-fa20-42aa-b677-4742df7ea3c7.mp4

Edit:
Had it happen on my laptop today. Saw this error in the console.
![routing error](https://user-images.githubusercontent.com/8922227/224447823-03d0754d-9a42-4ded-be53-2effb733d5fa.png)

Update: after further investigation I found that my laptop, which I am experiencing this issue on, includes these script tags which link to the same js files shown in the error in the source.
```html
<script src="/static/js/2.7cb8d6dc.chunk.js"></script><script src="/static/js/main.8b7dea01.chunk.js"></script>
```
Meanwhile, my PC, which the redirect is working on, shows these tags in the source for https://peterportal.org/.
```html
<script src="/static/js/2.15a50d8e.chunk.js"></script><script src="/static/js/main.dfccda19.chunk.js"></script>
```

At the moment, the js files linked in the former source code don't actually exist. Trying to view them takes you to a 404 page, explaining why the error says unexpected '<' at position 1. The files linked in the latter do, so an outdated index page is being served to my laptop which sounds like a caching issue.

Update:
Opening [https://peterportal.org/](https://peterportal.org/) in incognito mode instead works. Opening in non-incognito tab works after clearing cache on my browser so this is definitely a caching issue.

Update:
May want to look into these [lines](https://github.com/icssc/peterportal-client/blob/759b9c77d09b8dd7fd3ada3947c8011d5adbf820/api/src/app.ts#L111-L117). I don't think they are behaving as they should since https://peterportal.org/ is still getting cached.

Update 3/18/2023:
I think I've discovered the issue here. When making a request to the index, https://peterportal.org/, express isn't handling with [these lines](https://github.com/icssc/peterportal-client/blob/759b9c77d09b8dd7fd3ada3947c8011d5adbf820/api/src/app.ts#L111-L117), it's handling it with [this line](https://github.com/icssc/peterportal-client/blob/ffa95b06a92682f819a68a077efd8b1d56ced0bf/api/src/app.ts#L77) which is not using nocache. Solution may be the move that line down below the `app.use(nocache())`. A possible downside to implementing this solution would be that we would stop caching __all__ static files, including images.

**Comment by aahei** [comment]
I did some investigations on this issue but could not reproduce the situation. 

This is what I tried.

To simulate the deployment environment, I built the front end which produced a `/build` folder, which unique `.js` and `.css` file names. Then, I copied the `build` folder under `/api`, the back-end server folder, so that Express could serve the front end as static files, which were expected to be cached. When I started the Express server and visited the index.html, the status code was 200 the first time, and 304 afterward, which means the page was cached. 

To reproduce the issue, which happens when a new version of the front end is deployed, I rebuilt the front end and got a new `/build` folder with different `.js` and `.css` file names. Then, I copied the new `build` folder to the back-end server folder, `api`. As the issue describes, now the Express server should not provide the new `index.html` and other `.js .css` files when I visit `index.html`, which causes the error. However, the actual behavior was that the new `index.html` was transmitted with a status code of 200, which means that, despite the cache, the up-to-date files would always be provided, as opposed to the description.

This situation was discussed in the meeting on 2023-04-28. The decision was that the task should become on hold.
